### ISSUE TITLE
Add a helper to sign API responses with self-authenticated JWTs

### ISSUE THREAD

**Issue by tbinna** [toolsplus/atlassian-connect-play#11]
When sending a response from a server API you would typically add a `AUTHORIZATION` header with a self-authenticated JWT such that subsequent API calls can use a refreshed JWT.

Consider adding a helper to sign any response.

Possible implementation:

```scala
trait JwtResponseHelper {

  def selfAuthenticationTokenGenerator: SelfAuthenticationTokenGenerator
  def hostLogger: HostContextLogger

  def selfAuthenticationErrorResult =
    InternalServerError("Failed to generate self-authenticated JWT")

  /** Signs the given result using a JWT.
    *
    * Sings the result by adding an AUTHORIZATION header of the format 'JWT $token'.
    *
    * @param result Result to sign
    * @param hostUser Atlassian host user associated with this operation
    * @return Either the signed result or a JWT signing error
    */
  def withJwtAuthorizationHeaders(result: Result)(
      implicit hostUser: AtlassianHostUser): Either[JwtSigningError, Result] = {
    selfAuthenticationTokenGenerator
      .createSelfAuthenticationToken(hostUser)
      .map { token =>
        result.withHeaders(AUTHORIZATION -> s"JWT $token")
      }
  }

  /** Signs the given result using a JWT.
    *
    * Sings the result by adding an AUTHORIZATION header of the format 'JWT $token'.
    *
    * This is a convenience function that logs an error and returns either the
    * default error result or any result given as error result.
    *
    * @param result Result to sign
    * @param onError Result to return if signing the token fails
    * @param hostUser Atlassian host user associated with this operation
    * @return Signed result or error result if signing failed
    */
  def withJwtAuthorizationHeaders(result: Result,
                                  onError: Option[Result] = None)(
      implicit hostUser: AtlassianHostUser): Result = {
    withJwtAuthorizationHeaders(result) match {
      case Right(token) => result.withHeaders(AUTHORIZATION -> s"JWT $token")
      case Left(error) =>
        hostLogger.error(
          s"Failed to generate self-authenticated JWT: ${error.getMessage}")
        onError.getOrElse(selfAuthenticationErrorResult)
    }
  }

}
```

**Comment by tbinna** [comment]
In the example above it's probably better to remove the logger dependency as each class that mixes in this trait is forced to provide this logger.

**Comment by tbinna** [comment]
The header used in the response should be configurable, instead of using hard-coded `AUTHORIZATION` header. `AUTHORIZATION` should not be used anyways because it's not a valid HTTP response header.

**Comment by tbinna** [comment]
Fixed by #20
